File: ModSender/autonomy/RandomWalk.py
import time

import numpy as np

from ModSender.autonomy.Autonomous import Autonomous




class RandomWalk(Autonomous):

    def __init__(self, forward_force=0.3, min_distance=400, des_z=7):
        # Constants
        self.forward_force = forward_force
        self.min_distance = min_distance
        self.des_z = des_z
        self.time_backward = 1
        self.time_rotate = 2

        # Variable
        self.des_yaw = 0

        # current action
        self.current_action = 0

        # actions
        self.actions = [self._action_move_forward, self._action_move_backward, self._action_rotate, self._action_wait]

    def begin(self):
        self._restart_timer()


    def _choose_action(self, feedback):
        # Variables to make decisions
        distance = feedback[5]  # Distance from the sonar
        time_elapsed = self._time_elapsed()

        # ---------- Switch actions based on timer and distance -----------
        if self.current_action == 0 and distance < self.min_distance:
            self.current_action = 1  # Move backwards
            self._restart_timer()

        elif self.current_action == 1 and time_elapsed > self.time_backward:
            self.current_action = 2  # Rotate
            self._restart_timer()
        elif self.current_action == 2:
            self.current_action = 3
        elif self.current_action == 3 and time_elapsed > self.time_rotate:
            self.current_action = 0  # Move forward

        print("Current action: ", self.current_action)
        return self.actions[self.current_action]

    def execute(self, feedback):

        # Select the current action function
        action = self._choose_action(feedback)

        return action()

    def _restart_timer(self):
        self.time_elapse = time.time()

    def _time_elapsed(self):
        return time.time() - self.time_elapse

    def _action_move_forward(self):
        return self.forward_force, self.des_z, self.des_yaw

    def _action_move_backward(self):
        return -self.forward_force, self.des_z, self.des_yaw

    def _action_rotate(self):
        self.des_yaw += np.random.uniform(0, np.pi) + np.pi / 2
        return self.forward_force, self.des_z, self.des_yaw

    def _action_wait(self):
        return 0, self.des_z, self.des_yaw

File: randomwalk.cpp
#include "RandomWalk.h"

// Constructor to initialize member variables
RandomWalk::RandomWalk(float forward_force, int min_distance, int des_z):
    _forward_force(forward_force), _min_distance(min_distance), _des_z(des_z),
    _time_backward(1000), _time_rotate(2000), _des_yaw(0), _current_action(0)
{}

// Method to reset the timer
void RandomWalk::begin() {
    restart_timer();
}

// Method to execute actions based on feedback
void RandomWalk::execute(int distance_sensor, float &force, int &z, float &yaw) {
    choose_action(distance_sensor);  // Choose the next action
    // Execute the chosen action
    switch (_current_action) {
        case 0:
            action_move_forward(force, z, yaw);
            break;
        case 1:
            action_move_backward(force, z, yaw);
            break;
        case 2:
            action_rotate(force, z, yaw);
            break;
        case 3:
            action_wait(force, z, yaw);
            break;
    }
}

// Method to choose the next action based on feedback
void RandomWalk::choose_action(int distance_sensor) {
    int distance = distance_sensor;  // Get distance from feedback
    unsigned long elapsed_time  = time_elapsed();  // Get elapsed time

    // Choose the next action based on current action, distance, and elapsed time
    if (_current_action == 0 && distance < _min_distance) {
        _current_action = 1;
        restart_timer();
    } else if (_current_action == 1 && elapsed_time > _time_backward) {
        _current_action = 2;
        restart_timer();
    } else if (_current_action == 2) {
        _current_action = 3;
    } else if (_current_action == 3 && elapsed_time > _time_rotate) {
        _current_action = 0;
    }
}

// Method to reset the timer
void RandomWalk::restart_timer() {
    _time_elapse = millis();  // Get the current time in milliseconds
}

// Method to get the elapsed time since the timer was reset
unsigned long RandomWalk::time_elapsed() {
    return millis() - _time_elapse;  // Calculate and return elapsed time
}

// Action method to move forward
void RandomWalk::action_move_forward(float &force, int &z, float &yaw) {
    force = _forward_force;
    z = _des_z;
    yaw = _des_yaw;
}

// Action method to move backward
void RandomWalk::action_move_backward(float &force, int &z, float &yaw) {
    force = -_forward_force;
    z = _des_z;
    yaw = _des_yaw;
}

// Action method to rotate
void RandomWalk::action_rotate(float &force, int &z, float &yaw) {
    _des_yaw += random(0, 314) / 100.0 + PI / 2;  // Generate a random yaw value
    force = _forward_force;
    z = _des_z;
    yaw = _des_yaw;
}

// Action method to wait
void RandomWalk::action_wait(float &force, int &z, float &yaw) {
    force = 0;
    z = _des_z;
    yaw = _des_yaw;
}


// Setter method for forward_force
void RandomWalk::setForwardForce(float forward_force) {
    _forward_force = forward_force;
}

// Setter method for min_distance
void RandomWalk::setMinDistance(int min_distance) {
    _min_distance = min_distance;
}

// Setter method for des_z
void RandomWalk::setDesZ(int des_z) {
    _des_z = des_z;
}


